### ISSUE TITLE
internal error: extractDecl when selectively reexporting data instance

### ISSUE THREAD

**Issue by watashi** [haskell/haddock#979]
## Env

* ghc and haddock built from ghc master

```
$ uname -a
Linux localhost 4.18.8-arch1-1-ARCH #1 SMP PREEMPT Sat Sep 15 20:34:48 UTC 2018 x86_64 GNU/Linux
$ cabal --version
cabal-install version 2.5.0.0
compiled using version 2.5.0.0 of the Cabal library 
```

## Repro

Here is the minimal repro I constructed, with two packages where pkg2 depends on pkg1:

```haskell
$ cat pkg1/A.hs
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeFamilies #-}
module A
  ( C(..)  -- [1]
  , T(TI)
  , error
  ) where

class C a where
  data T a

instance C Int where
  data T Int = TI { ti :: Int }

$ cat pkg2/B.hs
module B
  ( module A
  ) where

import A hiding (error) -- [2]

$ cabal sandbox init
$ cabal add-source pkg1 pkg2
$ cabal install pkg1 pkg2 --with-ghc ~/ghc/inplace/bin/ghc-stage2 --with-haddock ~/ghc/inplace/bin/haddock --haddock-html --enable-documentation --reinstall
...
haddock: internal error: internal: extractDecl
CallStack (from HasCallStack):
  error, called at src/Haddock/Interface/Create.hs:1116:12 in haddock-api-2.22.0-inplace:Haddock.Interface.Create
...
```

## Issue

I looked into this a little bit and this is what the callstack looks like:

```haskell
mkExportItems _...
  Just [(module A, [C{T}, T{TI}])] -- exportList: without @C(..)@ at [1], this will be [T{T}, T{TI}]
  [T{TI}, C{T}] -- allExports; without @C(..)@ at [1], this will be [T{T, TI}] and haddock works
  _ _
-- we go to the last case of 'lookupExport' (haddock works without @hiding (error)@ at [2]
availExportItem _... T{TI}
-- the decl of 'T' is from another package: 'hiDecl' returns @data family T a :: Type@ (haddock works otherwise)
availExportDecl T{TI} (data family T a :: Type) (_, [(TI, _)])
-- @availExportsDecl avail@ is 'False', so we handle the sub one 'TI'
extractDecl [] TI (data family T a :: Type)
-- name doesn't exist in the decl returned by 
```

`availExportsDecl` depends on the invariants stated in `Avail` that if the type or class is itself to be in scope, it must be *first* in this list. But this invariant doesn't quite hold for data instance above?

Not sure what's the best way to fix this, we can at least suppress this error by intentionally ignoring this case.

**Comment by harpocrates** [comment]
I'm having a tough time replicating this. What GHC commit were you using?

**Comment by harpocrates** [comment]
OK, I can replicate now. This looks like a GHC bug; check out the avails reported by `tcg_exports`:

  * for `A`, you get `[error, C{C, T;}, T{T, TI;}]`
  * for `B`, you get `[T{TI;}, C{C, T;}]`

I'll see if I can diagnose the issue in GHC. If not, you're right we should at least patch Haddock (and leave this issue open as a reminder).

**Comment by harpocrates** [comment]
I've opened https://ghc.haskell.org/trac/ghc/ticket/16077 regarding the Avail invariant being violated.

**Comment by brandon-leapyear** [comment]
Hi! Getting a similar error over in [esqueleto](https://github.com/bitemyapp/esqueleto/issues/110)

**Comment by brandon-leapyear** [comment]
So it seems like on ghc-8.2.2, the export list is kept at just the module
```
-- group_
""

-- imports
[import (implicit) Prelude,
       import Database.Persist.Sql
           hiding ( BackendSpecificFilter,
                    BackendSpecificFilter,
                    Filter(..),
                    PersistQuery,
                    SelectOpt(..),
                    Update(..),
                    delete,
                    deleteWhereCount,
                    updateWhereCount,
                    selectList,
                    selectKeysList,
                    deleteCascadeWhere,
                    (=.),
                    (+=.),
                    (-=.),
                    (*=.),
                    (/=.),
                    (==.),
                    (!=.),
                    (<.),
                    (>.),
                    (<=.),
                    (>=.),
                    (<-.),
                    (/<-.),
                    (||.),
                    listToJSON,
                    mapToJSON,
                    getPersistMap,
                    limitOffsetOrder,
                    selectSource,
                    update,
                    count )],

-- mayExports
Just [module Database.Persist.Sql]

-- mayDocHeader
Just  Re-export "Database.Persist.Sql" without any clashes with\n @esqueleto@.
```
So it wasn't hitting the `DataFam` case in `extractDecl`

**Comment by Vlix** [comment]
Any news on this bug? Is there a way to work around it until it's fixed?

**Comment by harpocrates** [comment]
This GHC patch should fix the issue: https://gitlab.haskell.org/ghc/ghc/merge_requests/276. At the very least, it fixes the test case from https://github.com/haskell/haddock/issues/979#issue-390898190. No Haddock change will be required: this is just GHC not obeying one of its own invariants. 

In the meantime, it should be possible to work around this issue by replacing the problematic `module` re-exports with explicit export lists. I know this isn't a great state of affairs, but I'm not sure what more can be done for now. 😕

It would eventually be nice for Haddock to be more flexible around its error handling (for example: printing errors and exiting with an error code, but still producing a best effort at docs), but there are more pressing problems at the moment...

**Comment by harpocrates** [comment]
The GHC patch has been merged and cherry-picked to the 8.6 and 8.8 branches. This should mean that this will be fixed in `haddock` shipping with `ghc-8.6.4`

**Comment by MaxGabriel** [comment]
Are other people no longer seeing this when compiling with GHC 8.6.4? I'm still getting the error for e.g. classy-prelude-yesod-1.5.0, and that package is marked as broken because of this bug on the Nix 19.03 release for this bug. https://github.com/NixOS/nixpkgs-channels/blob/3a4ffdd38b56801ce616aa08791121d36769e884/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L3482

**Comment by normenmueller** [comment]
Updated to LTS 13.15 for GHC 8.6.4 but still get this error:

```bash
$ stack hoogle
No Hoogle database yet. Automatically building haddocks and hoogle database (use --no-setup to disable) ...
Yapper-0.0.0: unregistering
classy-prelude-yesod-1.5.0: configure
classy-prelude-yesod-1.5.0: build
classy-prelude-yesod-1.5.0: haddock
Progress 1/2

--  While building package classy-prelude-yesod-1.5.0 using:
      /Users/nrm/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.4 --builddir=.stack-work/dist/x86_64-osx/Cabal-2.4.0.1 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source --haddock-option=--quickjump
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/nrm/Sources/mdpm/projects/Yapper/.stack-work/logs/classy-prelude-yesod-1.5.0.log

    Configuring classy-prelude-yesod-1.5.0...
    Preprocessing library for classy-prelude-yesod-1.5.0..
    Building library for classy-prelude-yesod-1.5.0..
    [1 of 2] Compiling ClassyPrelude.Yesod ( src/ClassyPrelude/Yesod.hs, .stack-work/dist/x86_64-osx/Cabal-2.4.0.1/build/ClassyPrelude/Yesod.o )
    [2 of 2] Compiling Paths_classy_prelude_yesod ( .stack-work/dist/x86_64-osx/Cabal-2.4.0.1/build/autogen/Paths_classy_prelude_yesod.hs, .stack-work/dist/x86_64-osx/Cabal-2.4.0.1/build/Paths_classy_prelude_yesod.o )
    Preprocessing library for classy-prelude-yesod-1.5.0..
    Running Haddock on library for classy-prelude-yesod-1.5.0..
    Haddock coverage:
    haddock: internal error: internal: extractDecl
    CallStack (from HasCallStack):
      error, called at utils/haddock/haddock-api/src/Haddock/Interface/Create.hs:1116:12 in main:Haddock.Interface.Create
```

**Comment by 5outh** [comment]
I am still seeing this error with LTS 13.15 as well.

**Comment by skress** [comment]
The reason for haddock's error is a bug in GHC which was fixed (https://gitlab.haskell.org/ghc/ghc/merge_requests/276/commits) but not merged into the 8.6 branch.
 
So it seems like you need to wait for GHC 8.8 ... or you could checkout GHC's 8.6 branch cherry-pick 2a431640d1 and then compile your fixed GHC ... but that will take a lot of time ...
